fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/bow/BowExtensions.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.item.v1.bow;
+
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.projectile.PersistentProjectileEntity;
+import net.minecraft.item.BowItem;
+import net.minecraft.item.ItemStack;
+
+/**
+ * An interface to implement for all custom bows in fabric. <br>
+ * Note: This is meant to be used on a BowItem class, the functionality will not work otherwise.
+ *
+ * @see SimpleBowItem
+ */
+public interface BowExtensions {
+	/**
+	 * In this method you can modify the behavior of arrows shot from your custom bow. Applies all of the vanilla arrow modifiers first.
+	 *
+	 * @param bowStack                   The ItemStack for the Bow Item
+	 * @param arrowStack                 The ItemStack for the arrows
+	 * @param user                       The user of the bow
+	 * @param pullProgress          	 The pull progress of the bow from 0.0 to 1.0
+	 * @param persistentProjectileEntity The arrow entity to be spawned
+	 */
+	void onProjectileShot(ItemStack bowStack, ItemStack arrowStack, LivingEntity user, float pullProgress, PersistentProjectileEntity persistentProjectileEntity);
+
+	/**
+	 * Returns the pull progress of the bow between 0 and 1.
+	 *
+	 * @param useTicks The number of ticks the bow has been pulled.
+	 * @return The progress of the pull from 0.0f to 1.0f.
+	 */
+	default float getCustomPullProgress(int useTicks) {
```

Comment:
Hmm, can you offer item stack access in this method? [`onStoppedUsing`](https://github.com/FabricMC/fabric/pull/1210/files#diff-87cd612012ddfcde82094648318e7dc39dc9eeef80ba2c80c52583fb8ece0e32R46) passes the item stack afaik.

Reply:
It doesnt seem to, but it might be something useful, like with a quickcharge enchantment.

Reply:
hm looking into the minecraft code, id have to make some pretty nasty mixins to pass in the bow stack. but who cares about implementation details if it works 😎.